framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md

```diff
@@ -0,0 +1,54 @@
+# CoarsenBlockGenerator
+
+!syntax description /Mesh/CoarsenBlockGenerator
+
+## Overview
+
+This MeshGenerator object allows the user to coarsen one or more blocks in a mesh. It has been designed
+to work with previously refined meshes that were saved to a file without the mesh refinement pattern information. For example, refined meshes stored using [Exodus.md] or [VTK.md] file formats do not contain the information needed to recognize parent and children elements from uniform refinement. The heuristic can still work in the general case of non-uniform refinement.
```

Comment:
So the first thing I tried was changing your 2x2 block in `coarsen_quad.i` to 3x3 and I get the result I expected to get which is hanging nodes. I think that more often than not this generator will not work with meshes that were not generated through uniform block refinement. I would consider dropping this last sentence; if you don't want to drop it, then an example of a non-uniform case in which this will succeed might be good.

I think that ideally the generator should error if hanging nodes are generated

Reply:
the internal sideset examples are non-uniform refinement (not uniform over the whole domain, or even a whole block)

I can put the mesh diagnostics as a sub-generator if you d like

Reply:
> more often than not 

Absolutely in the infinite space of all meshes.
In the also infinite but smaller space of AMR-refined meshes for which we lost the adaptivity information, this is where it needs to shine

Reply:
Can you describe the use case for this? Previously you said

> sure but it also would not be very useful since we could only work with meshes refined with libmesh AMR

in reference to a coarsening generator that worked in the same way as the current refinement generator but now you are saying that the intent of your object is to work with AMR-refined meshes. But for your object we need to add an additional input file and insert an additional step of writing out the mesh to exodus or vtk format. Are there a lot of people doing AMR with third-party meshing software before loading the mesh into MOOSE?

Reply:
> I can put the mesh diagnostics as a sub-generator if you d like. This sub-generator can be set to error on the presence of non-conformalities. It will need to be switchable on/off as sometimes we will need two rounds of coarsening to get two separate refined regions coarsened

actually i dont think I can do mesh diagnostics as a sub-generator.
But I could move the code that detects non-conformality to a MeshDiagnosticsUtil.C and call it from the new generator? What do you think?